### PR TITLE
Tiered pricing and quantity Increments do not work with decimal inventory MAGETWO-86164 (both front-end and back-end)

### PR DESCRIPTION
**Pull request by vasilii-b**
## Description
All issues below are described in #12914 

#### Steps to reproduce:
1. Create new product/ edit existing one
2. Open **Advanced Inventory** pop-up  under  **Quantity**
3. Set **Qty Uses Decimals** to *Yes*
4. Set **Qty** value to decimal value
5. Set **Enable Qty Increments** to *Yes*
6. Set  **Qty Increments** a decimal value
7. Press DONE (may skip this step)
8. Open **Advanced Inventory** pop-up  under  **Quantity** (optional if 7 was skipped)
9. Set Set **Qty Uses Decimals** to *No*
10. Set **Qty** value to integer value
11. Set **Enable Qty Increments** to *No*
12. Press DONE or the X icon to close the button

#### Expected Result
Close the **Advanced Inventory** pop-up

#### Actual result
Pop-up not closed and no errors are visible

### [2] Product Tiered Pricing Issue

**Steps to reproduce**
1. Create Simple Product
2. Set Inventory to 100
3. Set "Qty Uses Decimals" in Advanced Inventory to "Yes"
4. Set Base Price to 3.99
5. Open Advanced Pricing tab
6. Set a Tiered Price for Qty of .5 and price of 2.99
7. Save

#### Steps to reproduce
1. Open Advanced Inventory
2. Set "Enable Qty Increments" to "Yes"
3. Set "Qty Increments" to ".5"
4. Save

#### Expected result
During each of the above scenarios (2,3), the product should save and allow tiered pricing and qty increments to work with decimal-based inventory

#### Actual result
Product does not save, admin user is given error stating "Please enter a valid number"

#### Steps to reproduce
1. Create Simple Product
2. Set Inventory to 100
3. Set "Qty Uses Decimals" in Advanced Inventory to "Yes"
4. Set Base Price to 3.99
5. Open Advanced Pricing tab
6. Set a Tiered Price for Qty of 1.5 and price of 2.99
7. Save
8. Open Advanced Pricing pop-up

#### Expected result
Tier price Qty is 1.5

#### Actual result
Tier price Qty is 1

#### Steps to reproduce
Use the same steps from [1] or [4]

#### Expected result
In cart popup, cart page, checkout cart summary product Qty appear with decimals (if applicable)

#### Actual result
In cart popup, cart page, checkout cart summary product Qty appear without decimals

### CODE REVIEW

**Comment by rogyar** [comment]
Please, fix the failed tests. Thanks

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=13359) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=13359) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: vasilii-b<br/>:x: Vasilii Burlacu<hr/>**Vasilii Burlacu** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

**Comment by vasilii-b** [comment]
@rogyar Could you please advice on the only remaining [failed test](https://travis-ci.org/magento/magento2/jobs/357752859) ? It says to reduce `__construct` method parameters to under 10, while it has 10 now.
Thank you!

**Comment by rogyar** [comment]
As an option, you can add the stockItemRepository to the helper class (`Magento\Bundle\Helper\Data`) instead of injecting it directly to the dataProvider.

**Comment by vasilii-b** [comment]
Hi @rogyar , can you please advice on the latest failed test ? I see failed tests has nothing related to the work I've done in scope of this PR. Should I still fix them ?

Thank you!

**Comment by PieterCappelle** [comment]
I'm following this thread, having same issue with one of our clients.

File: app/code/Magento/Bundle/Helper/Data.php [hunk 1]
```diff
@@ -17,15 +19,23 @@ class Data extends \Magento\Framework\App\Helper\AbstractHelper
      */
     protected $config;
 
+    /**
+     * @var StockItemRepositoryInterface
+     */
+    protected $stockItemRepository;
```
Comment (ishakhsuvarov) [inline app/code/Magento/Bundle/Helper/Data.php]: Please avoid creating new `private` methods and properties. Technical Direction of Magento favor composition over inheritance.

File: app/code/Magento/Bundle/Helper/Data.php [hunk 2]
```diff
@@ -40,4 +50,20 @@ public function getAllowedSelectionTypes()
 
         return isset($configData['allowed_selection_types']) ? $configData['allowed_selection_types'] : [];
     }
+
+    /**
+     * @param $productId
+     *
+     * @return bool
+     */
+    public function isProductQtyDecimal($productId)
```
Comment (ishakhsuvarov) [inline app/code/Magento/Bundle/Helper/Data.php]: Helper classes are considered to be an anti-pattern. It's best to extract this part in to a separate class with a proper single responsibility. 

Additionally, please consider the following:
* Type hinting for method arguments and return types are required.
* `$productId` variable name hints that it's a product, but the interface contract used below, seems to be a stock item id

File: app/code/Magento/CatalogInventory/Helper/ProductStockIsQtyDecimal.php [hunk 3]
```diff
@@ -0,0 +1,43 @@
+<?php
+
+namespace Magento\CatalogInventory\Helper;
```
Comment (ishakhsuvarov) [inline app/code/Magento/CatalogInventory/Helper/ProductStockIsQtyDecimal.php]: `Helper` namespace does not seem to be the best choice as it does not have any information on what it's related to. Usually it's used for classes which didn't find their place in the codebase.

File: app/code/Magento/CatalogInventory/Helper/ProductStockIsQtyDecimal.php [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+/**
+ * Class GetProductIsQtyDecimalService
+ */
+class ProductStockIsQtyDecimal
+{
+    /**
+     * @var StockItemRepositoryInterface
+     */
+    protected $stockItemRepository;
```
Comment (ishakhsuvarov) [inline app/code/Magento/CatalogInventory/Helper/ProductStockIsQtyDecimal.php]: Please don't use `protected`.

File: app/code/Magento/CatalogInventory/Helper/ProductStockIsQtyDecimal.php [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+    /**
+     * @param StockItemRepositoryInterface $stockItemRepository
+     *
+     * @return void
```
Comment (ishakhsuvarov) [inline app/code/Magento/CatalogInventory/Helper/ProductStockIsQtyDecimal.php]: `@return` annotation for the `__construct` is not required. Additionally, constructors usually return instance of the class they are constructing.

File: app/code/Magento/CatalogInventory/Helper/ProductStockIsQtyDecimal.php [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    {
+        $this->stockItemRepository = $stockItemRepository;
+    }
+
+    /**
+     * @param $productId
+     *
+     * @return bool
+     */
+    public function execute($productId)
```
Comment (ishakhsuvarov) [inline app/code/Magento/CatalogInventory/Helper/ProductStockIsQtyDecimal.php]: Please add type hinting for parameters and return type

File: app/code/Magento/CatalogInventory/Helper/ProductStockIsQtyDecimal.php [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    {
+        try {
+            $productStock = $this->stockItemRepository->get($productId);
+
+            return (bool) $productStock->getIsQtyDecimal();
+        } catch (NoSuchEntityException $e) {
+            return false;
```
Comment (ishakhsuvarov) [inline app/code/Magento/CatalogInventory/Helper/ProductStockIsQtyDecimal.php]: Please consider looking for another way to implement this, avoiding exception-managed business logic.

File: app/code/Magento/CatalogInventory/Helper/ProductStockIsQtyDecimal.php (hunk 6 above)
Comment (ishakhsuvarov) [inline app/code/Magento/CatalogInventory/Helper/ProductStockIsQtyDecimal.php]: `$productId` seems to be not the most descriptive name, as stock item repository get accepts stock item id

File: app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/General.php [hunk 8]
```diff
@@ -412,6 +420,24 @@ protected function formatPrice($value)
...
     }
 
+    /**
+     * Make price qty even int or decimal
+     *
+     * @param $priceQty
+     *
+     * @return int
+     */
+    protected function formatPriceQty($priceQty)
```
Comment (ishakhsuvarov) [inline app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/General.php]: It seems that this method may be `private`.

File: app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/General.php [hunk 9]
```diff
@@ -41,20 +42,27 @@ class General extends AbstractModifier
...
         ArrayManager $arrayManager,
-        AttributeRepositoryInterface $attributeRepository = null
+        AttributeRepositoryInterface $attributeRepository = null,
+        ProductStockIsQtyDecimal $isProductQtyDecimal
     ) {
         $this->locator = $locator;
         $this->arrayManager = $arrayManager;
         $this->attributeRepository = $attributeRepository
             ?: \Magento\Framework\App\ObjectManager::getInstance()->get(AttributeRepositoryInterface::class);
+        $this->isProductQtyDecimal = $isProductQtyDecimal;
```
Comment (ishakhsuvarov) [inline app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/General.php]: It makes sense either to add new dependency in a Backwards Compatible way or refactor the one above to be BiC. Please align these.

**Review by ishakhsuvarov — CHANGES_REQUESTED** [review]

**Comment by ishakhsuvarov** [comment]
> I see failed tests has nothing related to the work I've done in scope of this PR. Should I still fix them ?

This fail relates to the work you have done indirectly. A new dependency added to the class triggered the failure. Usually high object coupling numbers indicates that class has too much responsibilities.

The class, where the test failed, indeed has too much different capabilities to it. You may try refactoring it and extracting some parts out of it.

**Comment by sidolov** [comment]
Hi @vasilii-b , will you continue progress with this issue?

**Comment by rogyar** [comment]
@vasilii-b. I'm closing this issue due to inactivity. Feel free to reopen it when you decide to proceed. Thank you

**Comment by vasilii-b** [comment]
Hi @rogyar @ishakhsuvarov ,
I've fixed all latest failed tests as well as added requested changes.
Can you please have a look ?

**Comment by vasilii-b** [comment]
@magento-engcom-team give me test instance

**Comment by magento-engcom-team** [comment]
Hi @vasilii-b. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @vasilii-b, here is your [new Magento instance](http://ec2-34-228-235-121.compute-1.amazonaws.com/pr-13359/).
Admin access: [http://ec2-34-228-235-121.compute-1.amazonaws.com/pr-13359//admin](http://ec2-34-228-235-121.compute-1.amazonaws.com/pr-13359//admin)
Login: `admin` Password: `123123q`

**Comment by vasilii-b** [comment]
Hi @ishakhsuvarov, could you please review this PR once again ?
Thank you!

File: app/code/Magento/Bundle/Ui/DataProvider/Product/BundleDataProvider.php [hunk 10]
```diff
@@ -24,17 +31,19 @@ class BundleDataProvider extends ProductDataProvider
...
+     * @param \Magento\Ui\DataProvider\AddFieldToCollectionInterface[] $addFieldStrategies
+     * @param \Magento\Ui\DataProvider\AddFilterToCollectionInterface[] $addFilterStrategies
      */
     public function __construct(
         $name,
         $primaryFieldName,
         $requestFieldName,
         CollectionFactory $collectionFactory,
         Data $dataHelper,
+        StockStateInterface $stockState,
```
Comment (sidolov) [inline app/code/Magento/Bundle/Ui/DataProvider/Product/BundleDataProvider.php]: Please, inject new dependencies according to our [Backward Compatible Development Guide](https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/)

File: app/code/Magento/Bundle/Ui/DataProvider/Product/Form/Modifier/BundlePanel.php [hunk 11]
```diff
@@ -69,7 +69,9 @@ public function __construct(
     }
 
     /**
-     * {@inheritdoc}
```
Comment (sidolov) [inline app/code/Magento/Bundle/Ui/DataProvider/Product/Form/Modifier/BundlePanel.php:72]: Please, revert this changes. Just remove curly braces for static test.

File: app/code/Magento/Bundle/Ui/DataProvider/Product/Form/Modifier/BundlePrice.php [hunk 12]
```diff
@@ -41,7 +41,8 @@ public function __construct(
     }
 
     /**
-     * {@inheritdoc}
```
Comment (sidolov) [inline app/code/Magento/Bundle/Ui/DataProvider/Product/Form/Modifier/BundlePrice.php:44]: Please, revert this changes. Just remove curly braces for static test.

File: app/code/Magento/Bundle/Ui/DataProvider/Product/Form/Modifier/BundleWeight.php [hunk 13]
```diff
@@ -30,7 +30,8 @@ public function __construct(ArrayManager $arrayManager)
     }
 
     /**
-     * {@inheritdoc}
+     * @param array $meta
```
Comment (sidolov) [inline app/code/Magento/Bundle/Ui/DataProvider/Product/Form/Modifier/BundleWeight.php]: Please, revert this changes. Just remove curly braces for static test.

File: app/code/Magento/Bundle/Ui/DataProvider/Product/Form/Modifier/Composite.php [hunk 14]
```diff
@@ -67,7 +67,8 @@ public function __construct(
     }
 
     /**
-     * {@inheritdoc}
```
Comment (sidolov) [inline app/code/Magento/Bundle/Ui/DataProvider/Product/Form/Modifier/Composite.php:70]: Please, revert this changes. Just remove curly braces for static test.

File: app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/AdvancedPricing.php [hunk 15]
```diff
@@ -123,7 +130,8 @@ public function __construct(
         Data $directoryHelper,
         ArrayManager $arrayManager,
         $scopeName = '',
-        GroupSourceInterface $customerGroupSource = null
+        GroupSourceInterface $customerGroupSource = null,
+        StockStateInterface $stockState
```
Comment (sidolov) [inline app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/AdvancedPricing.php]: Please, inject new dependencies according to our [Backward Compatible Development Guide](https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/)

File: app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/General.php [hunk 16]
```diff
@@ -41,20 +43,28 @@ class General extends AbstractModifier
...
     /**
      * @param LocatorInterface $locator
      * @param ArrayManager $arrayManager
+     * @param StockStateInterface $stockState
      * @param AttributeRepositoryInterface|null $attributeRepository
      */
     public function __construct(
         LocatorInterface $locator,
         ArrayManager $arrayManager,
+        StockStateInterface $stockState,
```
Comment (sidolov) [inline app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/General.php]: Please, inject new dependencies according to our [Backward Compatible Development Guide](https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/)

File: app/code/Magento/CatalogInventory/Api/StockStateInterface.php [hunk 17]
```diff
@@ -70,4 +70,14 @@ public function suggestQty($productId, $qty, $scopeId = null);
...
+
+    /**
+     * Check whenever product qty is decimal
+     *
+     * @param string|int $productId
+     * @param string|int|null $scopeId
+     *
+     * @return bool
+     */
+    public function isStockQtyDecimal($productId, $scopeId = null): bool;
```
Comment (sidolov) [inline app/code/Magento/CatalogInventory/Api/StockStateInterface.php]: We are not allowed to add new methods to the existing interfaces according to our [Backward Compatible Development Guide](https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/). Consider creating a new interface with `isStockQtyDecimal ` method

**Review by sidolov — CHANGES_REQUESTED** [review]

**Comment by vasilii-b** [comment]
Hi @sidolov, can you please see my latest changes according your requests ? Thank you!

File: app/code/Magento/Bundle/Helper/Data.php [hunk 18]
```diff
@@ -5,6 +5,9 @@
  */
 namespace Magento\Bundle\Helper;
 
+use Magento\CatalogInventory\Api\StockItemRepositoryInterface;
```
Comment (sivaschenko) [inline app/code/Magento/Bundle/Helper/Data.php]: Please remove the unused use references.

File: app/code/Magento/Bundle/Ui/DataProvider/Product/BundleDataProvider.php [hunk 19]
```diff
@@ -24,17 +32,19 @@ class BundleDataProvider extends ProductDataProvider
      * @param string $requestFieldName
      * @param CollectionFactory $collectionFactory
      * @param Data $dataHelper
-     * @param \Magento\Ui\DataProvider\AddFieldToCollectionInterface[] $addFieldStrategies
-     * @param \Magento\Ui\DataProvider\AddFilterToCollectionInterface[] $addFilterStrategies
+     * @param StockStateInterface $stockState
      * @param array $meta
      * @param array $data
+     * @param \Magento\Ui\DataProvider\AddFieldToCollectionInterface[] $addFieldStrategies
+     * @param \Magento\Ui\DataProvider\AddFilterToCollectionInterface[] $addFilterStrategies
```
Comment (sivaschenko) [inline app/code/Magento/Bundle/Ui/DataProvider/Product/BundleDataProvider.php:38]: Please remove the duplicate parameter declaration

File: app/code/Magento/Bundle/Ui/DataProvider/Product/BundleDataProvider.php [hunk 20]
```diff
@@ -24,17 +32,19 @@ class BundleDataProvider extends ProductDataProvider
...
      */
     public function __construct(
         $name,
         $primaryFieldName,
         $requestFieldName,
         CollectionFactory $collectionFactory,
         Data $dataHelper,
+        StockStateInterface $stockState = null,
```
Comment (sivaschenko) [inline app/code/Magento/Bundle/Ui/DataProvider/Product/BundleDataProvider.php]: Please move the new added dependency to the end of parameters list

File: app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/General.php [hunk 21]
```diff
@@ -41,20 +44,28 @@ class General extends AbstractModifier
...
 
+    /**
+     * @var StockStateInterface
+     */
+    private $stockState;
+
     /**
      * @param LocatorInterface $locator
      * @param ArrayManager $arrayManager
+     * @param StockStateInterface $stockState
```
Comment (sivaschenko) [inline app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/General.php]: Docblock for parameters does not match the actual sequence of parameters

File: app/code/Magento/CatalogInventory/Api/StockQtyDecimalInterface.php [hunk 22]
```diff
@@ -0,0 +1,23 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
```
Comment (sivaschenko) [inline app/code/Magento/CatalogInventory/Api/StockQtyDecimalInterface.php:7]: Please enable the strict type mode with `declare(strict_types=1);` and specify strict types for parameters

File: app/code/Magento/CatalogInventory/Model/StockState.php [hunk 23]
```diff
@@ -156,4 +158,19 @@ public function checkQuoteItemQty($productId, $itemQty, $qtyToCheck, $origQty, $
...
         return $this->stockStateProvider->checkQuoteItemQty($stockItem, $itemQty, $qtyToCheck, $origQty);
     }
+
+    /**
+     * @param string|int $productId
+     * @param string|int|null $scopeId
+     *
+     * @return bool
+     */
+    public function isStockQtyDecimal($productId, $scopeId = null): bool
```
Comment (sivaschenko) [inline app/code/Magento/CatalogInventory/Model/StockState.php]: Please enable strict type mode with `declare(strict_types=1);`

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Thanks for collaboration. Please see the code review notes.

File: app/code/Magento/Bundle/Ui/DataProvider/Product/BundleDataProvider.php (hunk 19 above)
Comment (vasilii-b) [reply app/code/Magento/Bundle/Ui/DataProvider/Product/BundleDataProvider.php:38]: Sorry, that is not a duplicated declaration. They are different.

**Comment by vasilii-b** [comment]
@sivaschenko can you please take a look on added changes ? Thank you!

File: app/code/Magento/Bundle/Ui/DataProvider/Product/BundleDataProvider.php (hunk 19 above)
Comment (sivaschenko) [reply app/code/Magento/Bundle/Ui/DataProvider/Product/BundleDataProvider.php:38]: Sorry, my bad, you are absolutely right!

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Thanks for the updates. The code is fine, however, there is an issue with commits. It looks like you made commits under two different accounts, CLA is signed only for one of them, can you please squash your changes under the correct github account

**Comment by vasilii-b** [comment]
Ok @sivaschenko , I'll do it. Thank you!

**Comment by vasilii-b** [comment]
Hi @sivaschenko, all checks are passed now. Thank you!

**Review by sivaschenko — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sivaschenko, thank you for the review.
[ENGCOM-3577](https://jira.corp.magento.com/browse/ENGCOM-3577) has been created to process this Pull Request

File: app/code/Magento/CatalogInventory/Api/StockQtyDecimalInterface.php [hunk 24]
```diff
@@ -0,0 +1,24 @@
...
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\CatalogInventory\Api;
+
+/**
+ *  StockQtyDecimalInterface
+ */
+interface StockQtyDecimalInterface
```
Comment (sivaschenko) [inline app/code/Magento/CatalogInventory/Api/StockQtyDecimalInterface.php:13]: This interface looks more like SPI than API, it's better to place it in the Model namespace

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Hi @vasilii-b thanks for the contribution. Can you please cover the introduced functionality with MFTF/integration tests.

**Comment by sivaschenko** [comment]
@vasilii-b closing this pull request due to inactivity. Feel free to reopen it if you'd like to continue progress on it.

**Comment by vasilii-b** [comment]
@sivaschenko perfect! I'll delete the branch then as I'm full of played ping-pong here: changes are accepted then after a while comes someone else and request changes (and so a couple of times). Moreover, I've spotted that it is a common approach to "kindly force" for integration tests. I'm doing contribution and as far as I understand it - that's what I like to help with.
Seriously, I'm done with it since this PR has more than 1 year from its creation and still playing ping-pong. It's either accepted either deleted at all.

P.S. moving that service from API to SPI folder could be done on Magento side.
